app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java

```diff
@@ -0,0 +1,490 @@
+            throw new EntityNotFoundException(Account.class);
+        }
+        if (hibernateAccount.getStatus() == AccountStatus.DISABLED) {
+            throw new AccountDisabledException();
+        }
+
+        // Verify password
+        PasswordAlgorithm passwordAlgorithm = hibernateAccount.getPasswordAlgorithm();
+        String passwordHash = hibernateAccount.getPasswordHash();
+        if (passwordAlgorithm == null || StringUtils.isBlank(passwordHash)) {
+            LOG.error("Account " + hibernateAccount.getId() + " is enabled but has no password");
+            throw new EntityNotFoundException(Account.class);
+        }
+        try {
+            if (!passwordAlgorithm.checkHash(passwordHash, signIn.getPassword())) {
+                // To prevent enumeration attacks, if the password doesn't match, throw 404 account not found.
+                throw new EntityNotFoundException(Account.class);
+            }
+        } catch (InvalidKeyException | InvalidKeySpecException | NoSuchAlgorithmException ex) {
+            throw new BridgeServiceException("Error validating password: " + ex.getMessage(), ex);
+        }
+
+        // Unmarshall account
+        return unmarshallAccount(hibernateAccount);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Account constructAccount(Study study, String email, String password) {
+        // Set basic params from inputs.
+        GenericAccount account = new GenericAccount();
+        account.setStudyId(study.getStudyIdentifier());
+        account.setEmail(email);
+
+        // Hash password.
+        PasswordAlgorithm passwordAlgorithm = PasswordAlgorithm.DEFAULT_PASSWORD_ALGORITHM;
+        String passwordHash;
+        try {
+            passwordHash = passwordAlgorithm.generateHash(password);
+        } catch (InvalidKeyException | InvalidKeySpecException | NoSuchAlgorithmException ex) {
+            throw new BridgeServiceException("Error creating password: " + ex.getMessage(), ex);
+        }
+        account.setPasswordAlgorithm(passwordAlgorithm);
+        account.setPasswordHash(passwordHash);
+
+        // Generate health code.
+        HealthId healthId = healthCodeService.createMapping(study.getStudyIdentifier());
+        account.setHealthId(healthId);
+
+        return account;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String createAccount(Study study, Account account, boolean sendVerifyEmail) {
+        // Initial creation of account. Fill in basic initial parameters.
+        String accountId = BridgeUtils.generateGuid();
+
+        HibernateAccount hibernateAccount = marshallAccount(account);
+        hibernateAccount.setId(accountId);
+        hibernateAccount.setStudyId(study.getIdentifier());
+        hibernateAccount.setCreatedOn(DateUtils.getCurrentMillisFromEpoch());
+        hibernateAccount.setModifiedOn(DateUtils.getCurrentMillisFromEpoch());
+        hibernateAccount.setPasswordModifiedOn(DateUtils.getCurrentMillisFromEpoch());
```

Comment:
There is currently no code path that allows us to create an account with no password. (Except possibly by creating an account in the DB directly.) So it seemed appropriate to tag the passwordModifiedOn at initial account creation.